### PR TITLE
Add good practices page for secrets

### PR DESCRIPTION
**Pull request by shannonxtreme**
Umbrella issue: #4491
Original PR for overhauled Secrets concept end state: #34168

This PR updates the glossary term for Secrets and adds a new page for Secrets good practices. 

This is part of a larger change to overhaul our Secrets docs. 

The scope of work is defined in [this Google doc](https://docs.google.com/document/d/1VQ8UFux1HifjXo9rRwm2-mOoUOzmMF7zGf-C0BhurdQ/edit#heading=h.nq4idzs84lru).

Preview: https://deploy-preview-34291--kubernetes-io-main-staging.netlify.app/docs/concepts/security/secrets-good-practices/

/sig docs
/language en
/cc @sftim

### CODE REVIEW

**Comment by shannonxtreme** [comment]
@kubernetes/sig-security-pr-reviews and @kubernetes/sig-docs-pr-reviews there's info that's missing that I'm not too sure what to add (two sections with a TBD or TODO). I can also just remove those sections. PTAL and suggest changes/additions.

File: content/en/docs/concepts/configuration/secret.md
Comment (PushkarJ) [inline content/en/docs/concepts/configuration/secret.md:34]: IMO, this caution is rightly strongly worded and the replacement text is missing many important points that will be lost if they are not covered anywhere else. So we should consider keeping this paragraph as is, on this page because folks would have bookmarked this page and suddenly seeing this "caution" section lost would be a big surprise. 

Please let me if there is a reason I am missing (which is very likely the case) for removing this though.

File: content/en/docs/concepts/configuration/secret.md
Comment (PushkarJ) [inline content/en/docs/concepts/configuration/secret.md:1240]: This can be covered after PR is merged, but it would be good to clarify that "tmpfs" == "in-memory storage" != "stored at rest"

File: content/en/docs/concepts/configuration/secret.md
Comment (PushkarJ) [inline content/en/docs/concepts/configuration/secret.md:1254]:
```suggestion
Any privileged containers on a node can access all Secrets used
```

File: content/en/docs/concepts/security/secrets-good-practices.md
Comment (PushkarJ) [inline content/en/docs/concepts/security/secrets-good-practices.md:49]:
```suggestion
Using short lived secrets and implementing audit rules that alert on specific events like concurrent reading of secrets by a single user account, are possible ways to detect or limit the impact caused by secret data being exposed intentionally or unintentionally by a user who has access to a pod with secret data.
```

File: content/en/docs/concepts/security/secrets-good-practices.md
Comment (PushkarJ) [inline content/en/docs/concepts/security/secrets-good-practices.md:57]:
```suggestion
If you have multiple `etcd` instances, configure encrypted SSL/TLS communication between
the instances to protect the secret data in transit.
```

File: content/en/docs/concepts/security/secrets-good-practices.md
Comment (PushkarJ) [inline content/en/docs/concepts/security/secrets-good-practices.md:63]:
```suggestion
You can use a third-party Secrets store providers to keep your Secrets outside
```

File: content/en/docs/concepts/security/secrets-good-practices.md
Comment (PushkarJ) [inline content/en/docs/concepts/security/secrets-good-practices.md:88]:
```suggestion
After your application has read the contents of a Secret, it should not log the
Secret data in clear, and should not transmit the Secret data
```

File: content/en/docs/concepts/security/secrets-good-practices.md
Comment (PushkarJ) [inline content/en/docs/concepts/security/secrets-good-practices.md:93]: >Base64 encoding is _not_ an encryption method, it provides no additional confidentiality over plain text.

I feel this sentence deserves it on _Caution / Warning _ block

File: content/en/docs/concepts/security/secrets-good-practices.md
Comment (PushkarJ) [inline content/en/docs/concepts/security/secrets-good-practices.md:74]: Honestly, we should not provide guidance for this usecase IMO because this is out of scope of kubernetes project.

**Review by PushkarJ — CHANGES_REQUESTED** [review]
Added some feedback from perspective of end users who are fighting to justify why their external secrets provider solutions are worth investing time and money on. 

Overall really love the changes and refactoring of best practices in a dedicated page!

File: content/en/docs/concepts/configuration/secret.md
Comment (shannonxtreme) [reply content/en/docs/concepts/configuration/secret.md:1269]: I'm going to keep as might because we don't know the exact privileges of the containers (the intent wasnt "containers with privileged: true" -- is that what this implies?

Comment (PushkarJ) [reply content/en/docs/concepts/configuration/secret.md:1269]: Yeah atleast I read it as "privileged: true" :)

**Comment by PushkarJ** [comment]
/lgtm

(Overall looks great @shannonxtreme. Since this is a fairly important update would love another set of eyes from someone else in SIG Auth or Security too, before merging)

File: content/en/docs/concepts/configuration/secret.md
Comment (ritazh) [reply content/en/docs/concepts/configuration/secret.md:1269]: Might be good to explicitly call this out

```
When a container runs with `privileged: true`, it becomes a privileged container that can access all secrets on the host.
```

File: content/en/docs/concepts/configuration/secret.md
Comment (ritazh) [inline content/en/docs/concepts/configuration/secret.md:1274]: nit: s/Good/Recommended

Comment (ritazh) [reply content/en/docs/concepts/configuration/secret.md:1274]: nvm i see the other doc is also "good practices"

**Comment by ritazh** [comment]
/lgtm

Few minor comments to address, otherwise lgtm. Thanks for adding this!

File: content/en/docs/concepts/configuration/secret.md
Comment (tengqm) [inline content/en/docs/concepts/configuration/secret.md:1253]: What does this reformatting mean to localization teams? It means they have to compare each and every line of the text because that is how git diff works. As it is shown here, the whole paragraph looks rewritten -- 5 lines removed and 4 lines added. The localization team has to check carefully which word has been changed. The net effect is the   whole paragraph has to be retranslated.

Do we really want to do this, with a compelling reason? Are these changes bring us any value?

Please stop massaging the text in this way. There are tons of topics we have not covered in the docs and there are several hundred open issues to be closed. Could you look into some of those issues instead of reediting the existing content in this way?

Comment (shannonxtreme) [reply content/en/docs/concepts/configuration/secret.md:1253]: You're right, I had this PR out for review before our discussion in Slack and in other PRs about larger changes like this. I'll revert some of the tangential changes that aren't strictly related to this PR and split the commits for the new page to make it easy to see where content was changed vs just moved.

Comment (shannonxtreme) [reply content/en/docs/concepts/configuration/secret.md]: @tengqm reverted most changes to existing Secrets concept (except for one clarity change). Split the creation of the new page into multiple commits so that the diffs are clear about what changed vs what moved. 

@PushkarJ @aramase @ritazh the content changes are identical to the previous state, nothing new to review :)

**Comment by shannonxtreme** [comment]
Looking for reviews and feedback from @kubernetes/sig-docs-pr-reviews so we can get this unstuck. 

@tengqm I'd also appreciate another review from you to see if the split commits help to parse the diff.

File: content/en/docs/concepts/security/secrets-good-practices.md
Comment (tengqm) [inline content/en/docs/concepts/security/secrets-good-practices.md:6]: We can add a `weight` to this page

File: content/en/docs/concepts/security/secrets-good-practices.md
Comment (tengqm) [inline content/en/docs/concepts/security/secrets-good-practices.md:33]:
```suggestion
consider the following guidelines for `Secret` objects. You should also follow the other guidelines in
[RBAC good practices](/docs/concepts/security/rbac-good-practices).
```

File: content/en/docs/concepts/security/secrets-good-practices.md
Comment (tengqm) [inline content/en/docs/concepts/security/secrets-good-practices.md:37]: This doesn't warrant a bullet by itself?

File: content/en/docs/concepts/security/secrets-good-practices.md
Comment (tengqm) [inline content/en/docs/concepts/security/secrets-good-practices.md:64]: Please always keep a link on the same line so that our tools can detect dangling links.

```suggestion
your cluster. The [Kubernetes Secrets Store CSI Driver](https://secrets-store-csi-driver.sigs.k8s.io/)
is a DaemonSet that lets
```

File: content/en/docs/concepts/security/secrets-good-practices.md
Comment (tengqm) [inline content/en/docs/concepts/security/secrets-good-practices.md:69]:
```suggestion
For a list of supported providers, refer to
[Providers for the Secret Store CSI Driver](https://secrets-store-csi-driver.sigs.k8s.io/concepts.html#provider-for-the-secrets-store-csi-driver).
```

File: content/en/docs/reference/glossary/secret.md
Comment (tengqm) [inline content/en/docs/reference/glossary/secret.md:21]:
```suggestion
are stored unencrypted by default, but can be configured to be
[encrypted at rest](/docs/tasks/administer-cluster/encrypt-data/#ensure-all-secrets-are-encrypted).
```

File: content/en/docs/concepts/security/secrets-good-practices.md
Comment (shannonxtreme) [reply content/en/docs/concepts/security/secrets-good-practices.md:6]: I don't think I properly understand how weights work, does the weight of adjacent pages also need to change to reorder in the TOC?

File: content/en/docs/reference/glossary/secret.md
Comment (shannonxtreme) [reply content/en/docs/reference/glossary/secret.md:21]: There must be a VS Code formatter setting that I'm missing that'll keep links on one line

File: content/en/docs/concepts/security/secrets-good-practices.md
Comment (tengqm) [reply content/en/docs/concepts/security/secrets-good-practices.md:6]: Explicit weight would be desirable. By default, Hugo sort pages by name.

**Comment by Sea-n** [comment]
Unhold since [the previous concern](https://github.com/kubernetes/website/pull/34291#pullrequestreview-1051795669) is solved.
/unhold

It would be great if you squash the commits by yourself.
/label tide/merge-method-squash

**Comment by shannonxtreme** [comment]
Will do Sean, was waiting till the very end to consolidate my commits 😁 thank you

**Comment by shannonxtreme** [comment]
Based on my conversation in Slack, I'd prefer to keep all these commits separate to split my changes for easier localization effort

/remove-label tide/merge-method-squash

**Comment by sftim** [comment]
I agree - these commits each represent a logically separate change.

File: content/en/docs/concepts/configuration/secret.md
Comment (sftim) [inline content/en/docs/concepts/configuration/secret.md:1288]: (nit) @shannonxtreme I'd move this link into the _What's next_ section.

**Comment by sftim** [comment]
Tech reviewers, is there any benefit to introducing / recommending the `kubernetes.io/enforce-mountable-secrets` annotation (eg on a `default` ServiceAccount).

If we do, we should link to https://kubernetes.io/docs/reference/labels-annotations-taints/#enforce-mountable-secrets

**Comment by shannonxtreme** [comment]
Wtf, it changed it to three commits again

**Comment by shannonxtreme** [comment]
@aramase / @ritazh / @PushkarJ could you pls LMK your thoughts on Tim's question:

> Tech reviewers, is there any benefit to introducing / recommending the `kubernetes.io/enforce-mountable-secrets` annotation (eg on a `default` ServiceAccount).
> 
> If we do, we should link to https://kubernetes.io/docs/reference/labels-annotations-taints/#enforce-mountable-secrets

File: content/en/docs/concepts/security/secrets-good-practices.md
Comment (reylejano) [inline content/en/docs/concepts/security/secrets-good-practices.md:53]: typo?
```suggestion
Consider wiping or shredding the durable storage used by `etcd` once it is
```

Comment (shannonxtreme) [reply content/en/docs/concepts/security/secrets-good-practices.md:53]: I can't believe I did this, thank you Rey

**Comment by shannonxtreme** [comment]
Ready for approval. note that this PR was already LGTM'd and folks in SIG docs have also provided feedback. 

Mentioning @zacharysarah @mtardy who are this week's PR wranglers

File: content/en/docs/reference/glossary/secret.md
Comment (SergeyKanzhelev) [inline content/en/docs/reference/glossary/secret.md:16]: do you need so much more? =)

Comment (shannonxtreme) [reply content/en/docs/reference/glossary/secret.md:16]: Lol good catch, thank you

**Comment by sftim** [comment]
The PR description implies this is still a draft - is that right?

**Comment by sftim** [comment]
@shannonxtreme there's no Netlify preview for this. Could you rebase against _main_ or make some other change to tickle Netlify and wake it up?

File: content/en/docs/concepts/security/secrets-good-practices.md
Comment (sftim) [inline content/en/docs/concepts/security/secrets-good-practices.md:37]:
```suggestion
- For components, restrict **watch** or **list** access to only the most privileged,
  system-level components.
  Only grant **get** access for Secrets where this is justified by the component's normal
  behavior.
- For human users, restrict access to **get**, **watch** or **list** Secrets. Consider
  using third party authorization mechanisms if you need to implement fine-grained
  restrictions (for example: allowing admins to read any Secret, except for Secrets
  that you annotated as `confidential.example/reads-prohibited`).
```

File: content/en/docs/concepts/security/secrets-good-practices.md
Comment (sftim) [inline content/en/docs/concepts/security/secrets-good-practices.md:39]:
```suggestion

{{< caution >}}
Access to **list** Secrets implies access to fetch the contents of those Secrets.
{{< /caution >}}

```

File: content/en/docs/concepts/security/secrets-good-practices.md
Comment (sftim) [inline content/en/docs/concepts/security/secrets-good-practices.md:65]:
```suggestion
You can use third-party secrets store providers to keep your confidential data
outside your cluster, and then enable Pods to access that secret information.
The [Kubernetes Secrets Store CSI Driver](https://secrets-store-csi-driver.sigs.k8s.io/)
is a DaemonSet that lets the kubelet retrieve secret information from your
external store(s), and present that secret information as a volume mounted into
specific Pods that you authorise to access that data.
```
?

File: content/en/docs/concepts/security/secrets-good-practices.md
Comment (sftim) [inline content/en/docs/concepts/security/secrets-good-practices.md:91]:
```suggestion
repository means the confidential data is available to everyone who
can read the manifest.
```

**Review by sftim — COMMENTED** [review]
My thoughts

**Comment by kbhawkey** [comment]
👀

**Comment by shannonxtreme** [comment]
Avoiding my computer this weekend but I'll implement and request review on Tuesday!

**Comment by kbhawkey** [comment]
Page previews:
https://deploy-preview-34291--kubernetes-io-main-staging.netlify.app/docs/concepts/security/secrets-good-practices/
https://deploy-preview-34291--kubernetes-io-main-staging.netlify.app/docs/concepts/configuration/secret/
https://deploy-preview-34291--kubernetes-io-main-staging.netlify.app/docs/concepts/security/multi-tenancy/

File: content/en/docs/concepts/security/secrets-good-practices.md
Comment (kbhawkey) [reply content/en/docs/concepts/security/secrets-good-practices.md:65]: @shannonxtreme , what do you think about the suggestion?

File: content/en/docs/concepts/security/secrets-good-practices.md
Comment (kbhawkey) [inline content/en/docs/concepts/security/secrets-good-practices.md:71]: Can you provide a one line sentence to introduce the 'Developers' section (what is this section about)?

**Comment by kbhawkey** [comment]
Hello @shannonxtreme .
Generally, the changes look good. A couple of the pages contain minor updates. The 'Recommendations' sections in the Secrets concepts page are moved to a new page.
For some reason, the Note on the page is not styled correctly (https://deploy-preview-34291--kubernetes-io-main-staging.netlify.app/docs/concepts/security/secrets-good-practices/#configure-access-to-external-secrets).

The changes look ready to merge. Review the recent suggestions.
/lgtm

**Comment by shannonxtreme** [comment]
@kbhawkey changes implemented!

**Comment by shannonxtreme** [comment]
What does the merge-method-squash label do? Ideally I'd like to keep the commits separate because they're intended to show layers to make it easier on localization teams. Sorry if this is too late 😅

**Comment by sftim** [comment]
I agree, these commits are logically separate.
/remove-label tide/merge-method-squash

**Comment by sftim** [comment]
You could squash this down more though, @shannonxtreme 

Commit d448992930a14becb5826d84dfc5839c65b4ecde for example seems to be a no-op. Could you get this down to somewhere between four and nine commits, something like that? Eighteen feels like too many.

**Comment by shannonxtreme** [comment]
@sftim I'll give it a go! I don't know if I can cut half the commits, but I'll try and merge commits that won't cause confusion ☺️

**Comment by shannonxtreme** [comment]
Hehehe @sftim / @kbhawkey it's down to 7!!

**Comment by kbhawkey** [comment]
The `ko` label was applied 👀

**Comment by shannonxtreme** [comment]
It's because I messed up my rebase earlier and fixed it right after @kbhawkey 😭 there's no Korean language stuff in my commits I swear

**Comment by shannonxtreme** [comment]
So close I can taste it 🥵

File: content/en/docs/concepts/configuration/secret.md
Comment (sftim) [reply content/en/docs/concepts/configuration/secret.md]: Slightly related: if you turn on `NodeSwap` _or_ run the kubelet with the otherwise-unsupported use of a local pagefile, tmpfs data _can_ be stored at rest.

(consider not enabling swap if this behavior isn't what you want; otherwise, encrypt your page files, folks)

File: content/en/docs/concepts/configuration/secret.md
Comment (sftim) [inline content/en/docs/concepts/configuration/secret.md:1291]: How about putting this link, _and_ a summary of the most salient points, into the caution callout that appears much earlier in the page. See line 32.

(and keep these lines as-is)

File: content/en/docs/concepts/security/secrets-good-practices.md
Comment (reylejano) [inline content/en/docs/concepts/security/secrets-good-practices.md:61]: non-blocking and for the future, we can provide strategies on using short-lived Secrets

**Comment by reylejano** [comment]
Thank you for the changes and overall a good addition, lgtm
I'll wait for a reply to sftim's [suggestion](https://github.com/kubernetes/website/pull/34291/files#r980418533) and officially add the lgtm label when it's resolved

File: content/en/docs/concepts/configuration/secret.md
Comment (sftim) [reply content/en/docs/concepts/configuration/secret.md:1291]: (@PushkarJ - this is a tweak on your idea, see what you think)

Comment (shannonxtreme) [reply content/en/docs/concepts/configuration/secret.md:1292]: Did so!

Comment (PushkarJ) [reply content/en/docs/concepts/configuration/secret.md:1292]: That looks good to me @sftim

**Comment by sftim** [comment]
Congratulations. This PR is a nice improvement.

**Pull request merged**
